Frameworks/UIKit/StarboardXaml/DisplayTexture.mm

```diff
@@ -0,0 +1,213 @@
+        // Clear the cache if the texture is nullptr
+        _displayTextureCache.erase(img);
+    } else {
+        // Cache the texture
+        _displayTextureCache[img] = texture;
+    }
+    _displayTextureCacheLock.unlock();
+}
+}
+
+DisplayTexture::DisplayTexture(int width, int height) {
+    _xamlImage = ::CreateWritableBitmap(width, height);
+}
+
+std::shared_ptr<IDisplayTexture> DisplayTexture::GetForCGImage(CGImageRef img) {
+    CGImageRetain(img);
+    auto cleanup = wil::ScopeExit([img] { CGImageRelease(img); });
+
+    // If the image has a backing texture, use it
+    std::shared_ptr<IDisplayTexture> texture = img->Backing()->GetDisplayTexture();
+    if (texture) {
+        return texture;
+    }
+
+    // If we have a cached texture for this image, use it
+    texture = ::CachedDisplayTextureForImage(img);
+    if (texture) {
+        return texture;
+    }
+
+    // Else, create and cache a new texture
+    // Note: Can't easily use make_shared here due to the private ctor being called
+    texture = std::shared_ptr<IDisplayTexture>(new DisplayTexture(img));
+    ::SetCachedDisplayTextureForImage(img, texture);
+    return texture;
+}
+
+void DisplayTexture::ClearCacheForCGImage(CGImageRef img) {
+    ::SetCachedDisplayTextureForImage(img, nullptr);
+}
+
+ComPtr<IInspectable> DisplayTexture::GetBitmapForCGImage(CGImageRef img) {
+    return GetForCGImage(img)->GetContent();
+}
+
+DisplayTexture::DisplayTexture(CGImageRef img) {
+    if (img->_imgType == CGImageTypePNG || img->_imgType == CGImageTypeJPEG) {
+        const void* data = NULL;
+        bool freeData = false;
+        int len = 0;
+        StrongId<NSData> fileData;
+
+        switch (img->_imgType) {
+            case CGImageTypePNG: {
+                CGPNGImageBacking* pngImg = (CGPNGImageBacking*)img->Backing();
+
+                if (pngImg->_fileName) {
+                    fileData.attach([[NSData alloc] initWithContentsOfFile:[NSString stringWithUTF8String:pngImg->_fileName]]);
+                    data = [fileData bytes];
+                    len = [fileData length];
+                } else {
+                    data = [(NSData*)pngImg->_data bytes];
+                    len = [(NSData*)pngImg->_data length];
+                }
+            } break;
+
+            case CGImageTypeJPEG: {
+                CGJPEGImageBacking* jpgImg = (CGJPEGImageBacking*)img->Backing();
+
+                if (jpgImg->_fileName) {
+                    fileData.attach([[NSData alloc] initWithContentsOfFile:[NSString stringWithUTF8String:jpgImg->_fileName]]);
+                    data = [fileData bytes];
+                    len = [fileData length];
+                } else {
+                    data = [(NSData*)jpgImg->_data bytes];
+                    len = [(NSData*)jpgImg->_data length];
+                }
+            } break;
+            default:
+                TraceError(TAG, L"Warning: unrecognized image format sent to DisplayTextureContent!");
```

Comment:
Never mind, it's getting removed in my change over to use WIC/D2D

---
In reply to: [86869562](https://github.com/Microsoft/WinObjC/pull/1318#discussion_r86869562) [](ancestors = 86869562)